rust/src/csv/reader.rs

```diff
@@ -200,6 +344,150 @@ impl Reader {
     }
 }
 
+/// CSV file reader builder
+pub struct ReaderBuilder {
+    /// Optional schema for the CSV file
+    ///
+    /// If the schema is not supplied, the reader will try to infer the schema
+    /// based on the CSV structure. The CSV should have headers for this to work.
+    schema: Option<Arc<Schema>>,
+    /// Whether the file has headers
+    has_headers: bool,
+    /// An optional column delimiter. Defauits to `b','`
+    delimiter: Option<u8>,
+    /// Optional maximum number of records to read during schema inference
+    ///
+    /// This is not required if a schema is provided, and currently defaults to
+    /// `MAX_INFERENCE_RECORDS`.
+    max_records: Option<usize>,
+    /// Batch size (number of records to load each time)
+    ///
+    /// The default batch size when using the `ReaderBuilder` is 1024 records
+    batch_size: usize,
+    /// Optional projection for which columns to load (zero-based column indices)
+    projection: Option<Vec<usize>>,
+}
+
+impl Default for ReaderBuilder {
+    fn default() -> ReaderBuilder {
+        ReaderBuilder {
+            schema: None,
+            has_headers: true,
+            delimiter: None,
+            max_records: None,
+            batch_size: 1024,
+            projection: None,
+        }
+    }
+}
+
+impl ReaderBuilder {
+    /// Create a new builder for configuring CSV parsing options.
+    ///
+    /// To convert a builder into a reader, call `Reader::from_builder`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// extern crate arrow;
+    ///
+    /// use arrow::csv;
+    /// use std::fs::File;
+    ///
+    /// fn example() -> Reader {
+    ///     let file = File::open("test/data/uk_cities_with_headers.csv").unwrap();
+    ///
+    ///     // create a builder, inferring the schema with the first 100 records
+    ///     let builder = csv::ReaderBuilder::new().infer_schema(Some(100));
+    ///
+    ///     let reader = csv::Reader::from_builder(file, builder);
+    ///
+    ///     reader
+    /// }
+    /// ```
+    pub fn new() -> ReaderBuilder {
+        ReaderBuilder::default()
+    }
+
+    /// Set the CSV file's schema
+    pub fn with_schema(&self, schema: Arc<Schema>) -> ReaderBuilder {
```

Comment:
Each builder method should take a `self` and return a `self`, e.g.,

```rust
pub fn with_schema(self, schema: Arc<Schema>) -> Self {
    self.schema = Some(schema)
    self
}
```

Reply:
Done